storage/src/adb/keyless.rs

```diff
@@ -0,0 +1,310 @@
+            },
+        )
+        .await?;
+
+        let values = VJournal::<E, V>::init(
+            context.with_label("values"),
+            VConfig {
+                partition: cfg.values_journal_partition,
+                compression: cfg.values_compression,
+                codec_config: cfg.values_codec_config,
+                buffer_pool: cfg.buffer_pool.clone(),
+                write_buffer: cfg.values_write_buffer,
+            },
+        )
+        .await?;
+
+        let locations = FJournal::init(
+            context.with_label("locations"),
+            FConfig {
+                partition: cfg.locations_journal_partition,
+                items_per_blob: cfg.locations_items_per_blob,
+                write_buffer: cfg.locations_write_buffer,
+                buffer_pool: cfg.buffer_pool,
+            },
+        )
+        .await?;
+
+        Ok(Self {
+            mmr,
+            values,
+            size: 0,
+            locations,
+            values_per_section: cfg.values_items_per_section.get(),
+            hasher,
+        })
+    }
+
+    /// Get the value at location `loc` in the database. Returns None if the location is valid but
+    /// does not correspond to an append.
+    pub async fn get(&self, loc: u64) -> Result<Option<V>, Error> {
+        match self.locations.read(loc).await {
+            Ok(op) => {
+                if let Operation::Append(offset) = op {
+                    let section = loc / self.values_per_section;
+                    let Some(v) = self.values.get(section, offset).await? else {
+                        panic!("didn't find value at location {loc} and offset {offset}");
+                    };
+
+                    Ok(Some(v))
+                } else {
+                    Ok(None)
+                }
+            }
+            Err(e) => Err(Error::Journal(e)),
+        }
+    }
+
+    /// Get the number of appends + commits that have been applied to the db.
+    pub fn size(&self) -> u64 {
+        self.size
+    }
+
+    /// Returns the section of the values journal where we are currently writing new items.
+    fn current_section(&self) -> u64 {
+        self.size / self.values_per_section
+    }
+
+    /// Return the oldest location that remains retrievable.
+    pub async fn oldest_retained_loc(&self) -> Result<Option<u64>, Error> {
+        self.locations
+            .oldest_retained_pos()
+            .await
+            .map_err(Error::Journal)
+    }
+
+    /// Append a value to the db, returning its location which can be used to retrieve it.
+    pub async fn append(&mut self, value: V) -> Result<u64, Error> {
+        let loc = self.size;
+        let section = self.current_section();
+        let (offset, _) = self.values.append(section, value).await?;
+        self.locations.append(Operation::Append(offset)).await?;
+        self.mmr
+            .add_batched(&mut self.hasher, &Operation::Append(offset).encode())
```

Comment:
To confirm my understanding (as always), the reason why we can't just ask the `mmr` for the item at position `n` (during a get) is that it just stores digests?

Reply:
(i.e. why we need to store `locations` too)